### DISCUSSION TITLE
Shorten Eldoc ?

### DISCUSSION THREAD

**Discussion by jdtsmith** [emacs-lsp/lsp-mode#3206] [Ideas]
lsp-mode's eldoc is fantastic in that it makes use of the current-argument highlighting, which I've only ever seen emacs-lisp mode do.  But it also places the entire `documentation` text in the echo area (truncated, but still quite long for many functions with my lsp server pyright).  

Since the hover response includes ranges, perhaps showing only the documentation text up to the next newline after the last mentioned range would be useful.  This would happen if `lsp-eldoc-render-all` is nil (as mine is).

**Comment by yyoncho** [comment]
Can you elaborate a bit more?

> Since the hover response includes ranges, perhaps showing only the documentation text up to the next newline after the last mentioned range would be useful.

The server response is pretty much free text.

**Comment by jdtsmith** [comment]
Looking closer, I see there is also `signatures[i].label`, which is what `signatures[i].parameters` indexes.  So I guess the simpler request is "add just the first line of documentation to the signature in eldoc".

**Reply by yyoncho** [reply]
That is not eldoc, this is lv. As a side note have you tried  `(setq lsp-signature-function 'lsp-signature-posframe)`

**Reply by jdtsmith** [reply]
Thanks.  I did try the posframe, but prefer a simpler echo-area signature.  Now that you mention it, eldoc doesn't seem to work on its timer, instead I get call syntax only when I open a paren in a function call.  But I see `lsp-eldoc-function` is wrapped around `eldoc-documentation-function` and calls `hover` in the end.  So not clear why the timer isn't working.  Is this expected?

**Reply by yyoncho** [reply]
Yes. In lsp there is hover and there signature help functions. We map hover to eldoc and signature help to lv/posframe. Signature help is displayed after pressing char that will trigger it.

**Reply by jdtsmith** [reply]
I see, the real issue is that I apparently never get hover/eldoc showing.  Is that something that has to be turned on? I do get `textDocument/signatureHelp` responses when logging LSP comms, but see no hover requests.  `lsp-eldoc-enable-hover` is `t`, and describe-session reports:

```
       |-[-] hoverProvider:
       |  `-[X] workDoneProgress: t
```

**Reply by yyoncho** [reply]
Can you test with lsp-start-plain.el after placing the cursor after a function/variable.

**Reply by jdtsmith** [reply]
Indeed starting from plain, it does work with my same lang-server.  Not sure what setting I have that would disable it.  I bisected my lsp-mode config, and found... nothing.  AKA, a restart with unchanged settings got it working again. I'll keep an eye out if it goes down again.
